### ISSUE TITLE
[Wine] SaveImage dialog exits Dolphin even if Cancel is pressed

### ISSUE THREAD

**Issue by brickviking** [dolphinsmalltalk/Dolphin#703]
When exiting Dolphin, a save-image requester turns up with "Yes", "No" and "Cancel" but clicking Cancel exits Dolphin instead of returning to Dolphin. No such issue under Vista, I get back to Dolphin fine there. I'll investigate, see what differences I can see between them.

**Comment by blairmcg** [comment]
Try evaluating the following and pressing the Cancel button:
```
MessageBox new beTaskModal; confirmOrCancel: 'Test'
```
If you get back #cancel, then this suggests that under Wine the PostQuitMessage/WM_QUIT cannot be denied, and terminates the application regardless. That may be hard to solve, but try putting a halt in DevelopmentSessionManager>>onQueryEndSession: in the onCancel block, and and then debug stepping through to see what happens. Under Windows the message loop will just continue if `queryEndSession` returns false. All the code is in Smalltalk and can be stepped through. If you don't get into the onCancel (or didn't get `#cancel` back abpve), then there is an issue with the message box implementation. You can try without the `beTaskModal` to see if that makes any difference.

A (rather unsatisfactory) workaround would be to use only the `confirm:` (yes/no) options when running under Wine to avoid offering a Cancel option that is ignored.

**Comment by brickviking** [comment]
Sorry that I took so long to get back to this, I've been having adventures in Minecraft-land. Anyhow, under Vista, I get back #cancel, under wine I get back #no. I'll trace through DevelopmentSessionManager like you suggest. I still get the same behaviour under 7.1.8

**Comment by brickviking** [comment]
In MessageBox>>result, I replaced the buttonStyle test with the following:
''' (self buttonStyle == #yesNo) 
'''
With merely the test for #yesNo, the system appears to work for the single case of Click on File, click on Exit Dolphin, dialog pops up, I click Cancel. So far, so good. I click on File, click on Exit Dolphin, dialog pops up, I click No, Dolphin quits. Again, does as it says on the tin.

I also rechecked that test you suggested, now I get #cancel when I click cancel. I'll tinker some more.

To me and my version of wine at least, it appears to be doing "The Right Thing™". Do you know anything further that would affect this?

**Comment by blairmcg** [comment]
Hmm, that Wine specific code in `MessageBox>>#result` looks suspect to me. The comment refers to "suppressible" message boxes having a bug, but then transforms the return value for all message boxes effectively destroying the support for the cancel button for normal message boxes as well. At the very least it should be performing that transformation only if `self isSuppressible` is true.

How does it behave if you debug through this:
```
MessageBox new beTaskModal; isSuppressible: true; uniqueId: GUID newUnique; confirmOrCancel: 'Test'
```
Is the comment correct? Do the suppressible message boxes not support the Yes and No buttons properly?

**Comment by brickviking** [comment]
What I did notice from your last code was a dialog that had a yes button, a no button, and a checkbox with a label "Don't show me this again". No Cancel button. Am I meant to be seeing one? If I remove the isSuppressible: stanza (and uniqueId), I get the Cancel button.

**Comment by blairmcg** [comment]
This is the result of the expression above on Windows 10:

![image](https://user-images.githubusercontent.com/15128107/62004739-e45ed200-b120-11e9-9dce-33d783f792ba.png)

From what you say its clear that Wine does not support the modes with yes and no buttons properly in its implementation of the ShMessageBoxCheck() API. Given that, I think the best solution is to move the current workaround to `MessageBox>>#showSuppressible` so that it is only applied when to suppressible message boxes. 

I've pushed a fix for this as commit 43d9269b (see branch blairmcg/703). I assume it will work, but I don't have a Wine setup to test that.

**Comment by brickviking** [comment]
Yup. I found the relevant function. It seems that SHDlgProc (in wine's source dlls/shlwapi/msgbox.c:198) only ends up using three values, which then is all the values that SHMessageBoxCheckA can use. I'm assuming that somewhere along the line, Dolphin ends up calling this function indirectly. That rather explains the funky behaviour I've been seeing with that dialog box in particular.
I'll do your patch in place and give that a try. Thanks for helping me nut this out.
